src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs

```diff
@@ -38,6 +38,10 @@ public static class EndpointRouteBuilderExtensions
         string pattern,
         RequestDelegate requestDelegate)
     {
+        if (requestDelegate.Method.ReturnType.IsGenericType)
```

Comment:
This doesn't seem right. This should be checking specifically for `Task<T>`.
```suggestion
        var returnType = f.Method.ReturnType;
        if (returnType is { IsGenericType: true } && returnType.GetGenericTypeDefinition() == typeof(Task<>))
```

Reply:
@davidfowl 

[RequestDelegate](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Abstractions/src/RequestDelegate.cs) use Task as ReturnType
`public delegate Task RequestDelegate(HttpContext context);`

if returnType.IsGenericType, it must be `Task<>`, right?

Reply:
Nope:

```C#
var f = MyClass<int> () => new(() => { });

Console.WriteLine(f.Method.ReturnType.IsGenericType);

class MyClass<T> : Task
{
    public MyClass(Action action) : base(action)
    {
    }
}
```

Reply:
Hmm, trying to get a better example (if possible) that shows this case with the overload.

Reply:
Here ya go:

```C#

Thing(() => new MyClass<int>());

void Thing(ReturnsATask del)
{
    Console.WriteLine(del.Method.ReturnType.IsGenericType);
}

delegate Task ReturnsATask();

class MyClass<T> : Task
{
    public MyClass() : base(() => { })
    {
    }
}

```

Reply:
Now that I've tried to reproduce this, I realized it doesn't seem to work at all. What is your test verifying exactly?

Reply:
OK scratch that, I can reproduce the problem now and the fix now. This is a tricky issue because I guess the problem only reproduces with method groups.

```C#

Task<int> Fails() => Task.FromResult(1);
MyClass<int> Fails1() => new MyClass<int>();

Mapper.Map(Fails1);

delegate Task ReturnsATask();

class MyClass<T> : Task
{
    public MyClass() : base(() => { })
    {
    }
}
```